app/workers/historical_data_deletion_worker.rb

```diff
@@ -0,0 +1,46 @@
+class HistoricalDataDeletionWorker < ApplicationWorker
+  def perform
+    # cascades matched content changes
+    delete_and_log(ContentChange) { ContentChange.where("created_at < ?", retention_period) }
+
+    # cascades matched messages
+    delete_and_log(Message) { Message.where("created_at < ?", retention_period) }
+
+    # cascades digest run subscribers
+    delete_and_log(DigestRun) { DigestRun.where("created_at < ?", retention_period) }
+
+    # deleting subscriptions must be done before deleting subscriber lists or subscribers
+    delete_and_log(Subscription) { Subscription.where("ended_at < ?", retention_period) }
+
+    delete_and_log(SubscriberList) { empty_subscriber_lists }
+
+    # restricts deletion if emails are present
+    delete_and_log(Subscriber) { Subscriber.where("deactivated_at < ?", retention_period) }
+  end
+
+private
+
+  def retention_period
+    @retention_period ||= 1.year.ago
+  end
+
+  def delete_and_log(model)
+    start_time = Time.zone.now
+    deleted_count = yield.delete_all
+    log_complete(deleted_count, Time.zone.now, start_time, model)
+  end
+
+  def log_complete(deleted, end_time, start_time, model)
+    seconds = (end_time - start_time).round(2)
+    model = model.to_s.underscore.humanize.downcase.pluralize
+    message = "Deleted #{deleted} #{model} in #{seconds} seconds"
+    logger.info(message)
+  end
+
+  def empty_subscriber_lists
+    SubscriberList
+      .where("subscriber_lists.created_at < ?", retention_period)
+      .joins("LEFT OUTER JOIN subscriptions ON subscriber_lists.id = subscriptions.subscriber_list_id")
+      .where("subscriptions.id IS NULL")
```

Comment:
You can do both of these clauses with more conventional ActiveRecord syntax: `left_outer_joins(:subscriptions)` and `where("subscriptions.id": nil)`

Reply:
Ahh of course active record has this covered, cheers!